core/src/test/java/kafka/server/share/DelayedShareFetchTest.java

```diff
@@ -385,6 +492,108 @@ public void testForceCompleteTriggersDelayedActionsQueue() {
         assertFalse(delayedShareFetch1.isCompleted());
         Mockito.verify(replicaManager, times(1)).addToActionQueue(any());
         Mockito.verify(replicaManager, times(0)).tryCompleteActions();
+        Mockito.verify(delayedShareFetch2, times(1)).releasePartitionLocks(any());
+    }
+
+    @Test
+    public void testCombineLogReadResponse() {
+        String groupId = "grp";
+        ReplicaManager replicaManager = mock(ReplicaManager.class);
+        TopicIdPartition tp0 = new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("foo", 0));
+        TopicIdPartition tp1 = new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("foo", 1));
+        Map<TopicIdPartition, Integer> partitionMaxBytes = new HashMap<>();
+        partitionMaxBytes.put(tp0, PARTITION_MAX_BYTES);
+        partitionMaxBytes.put(tp1, PARTITION_MAX_BYTES);
+
+        SharePartition sp0 = mock(SharePartition.class);
+        SharePartition sp1 = mock(SharePartition.class);
+
+        Map<TopicIdPartition, SharePartition> sharePartitions = new HashMap<>();
+        sharePartitions.put(tp0, sp0);
+        sharePartitions.put(tp1, sp1);
+
+        CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>> future = new CompletableFuture<>();
+        ShareFetchData shareFetchData = new ShareFetchData(
+            new FetchParams(ApiKeys.SHARE_FETCH.latestVersion(), FetchRequest.ORDINARY_CONSUMER_ID, -1, MAX_WAIT_MS,
+                1, 1024 * 1024, FetchIsolation.HIGH_WATERMARK, Optional.empty()), groupId, Uuid.randomUuid().toString(),
+            future, partitionMaxBytes, MAX_FETCH_RECORDS);
+
+        DelayedShareFetch delayedShareFetch = DelayedShareFetchBuilder.builder()
+            .withShareFetchData(shareFetchData)
+            .withReplicaManager(replicaManager)
+            .withSharePartitions(sharePartitions)
+            .build();
+
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new HashMap<>();
+        topicPartitionData.put(tp0, mock(FetchRequest.PartitionData.class));
+        topicPartitionData.put(tp1, mock(FetchRequest.PartitionData.class));
+
+        // Case 1 - logReadResponse contains tp0.
+
+        Map<TopicIdPartition, LogReadResult> logReadResponse = Collections.singletonMap(
+            tp0, mock(LogReadResult.class));
+        delayedShareFetch.updateLogReadResponse(logReadResponse);
+
+        doAnswer(invocation -> buildLogReadResult(Collections.singleton(tp1))).when(replicaManager).readFromLog(any(), any(), any(ReplicaQuota.class), anyBoolean());
+        Map<TopicIdPartition, LogReadResult> combinedLogReadResponse = delayedShareFetch.combineLogReadResponse(topicPartitionData);
```

Comment:
Could we change `combineLogReadResponse` to also take `partitionsAlreadyFetched`? This way, we can get rid of `delayedShareFetch.updateLogReadResponse`.